src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -203,10 +206,15 @@ public String url() {
      *  Note: to avoid deadlocks, when nesting locks we ALWAYS need to lock on the guard first, THEN the WorkflowRun.
      *  Synchronizing this helps ensure that fields are not mutated during a {@link #save()} operation, since that locks on the Run.
      */
-    private synchronized Object getMetadataGuard() {
+    private Object getMetadataGuard() {
         if (metadataGuard == null) {
-            metadataGuard = new Object();
+            synchronized (this) {
+                if (metadataGuard == null) {
+                    metadataGuard = new Object();
+                }
+            }
         }
```

Comment:
Just for future reference, @olamy reminded me that I was confused why we don't just initialize `metadataGuard` in `readResolve` so that we don't need any locking in this method. Would be worth looking into in a followup PR.